src/lib/storage/lz4_segment.cpp

```diff
@@ -120,27 +140,257 @@ std::vector<pmr_string> LZ4Segment<pmr_string>::decompress() const {
   return decompressed_strings;
 }
 
+template <typename T>
+void LZ4Segment<T>::_decompress_block(const size_t block_index, std::vector<T>& decompressed_data,
+                                      const size_t write_offset) const {
+  const auto decompressed_block_size = block_index + 1 != _lz4_blocks.size() ? _block_size : _last_block_size;
+  auto& compressed_block = _lz4_blocks[block_index];
+  const auto compressed_block_size = compressed_block.size();
+
+  int decompressed_result;
+  if (_dictionary.empty()) {
+    /**
+     * If the dictionary is empty, we either have only a single block or had not enough data for a dictionary.
+     * When decoding without a dictionary LZ4 needs a stream decode pointer (which would be used to decode the
+     * following blocks).
+     * A new decoder needs to be created for every block (in the case of multiple blocks being compressed without a
+     * dictionary) since the blocks were compressed independently.
+     */
+    LZ4_streamDecode_t lz4_stream_decoder;
+    auto lz4_stream_decoder_ptr = std::make_unique<LZ4_streamDecode_t>(lz4_stream_decoder);
+
+    decompressed_result = LZ4_decompress_safe_continue(lz4_stream_decoder_ptr.get(), compressed_block.data(),
+                                                       reinterpret_cast<char*>(decompressed_data.data()) + write_offset,
+                                                       static_cast<int>(compressed_block_size),
+                                                       static_cast<int>(decompressed_block_size));
+  } else {
+    decompressed_result = LZ4_decompress_safe_usingDict(
+        compressed_block.data(), reinterpret_cast<char*>(decompressed_data.data()) + write_offset,
+        static_cast<int>(compressed_block_size), static_cast<int>(decompressed_block_size), _dictionary.data(),
+        static_cast<int>(_dictionary.size()));
+  }
+
+  Assert(decompressed_result > 0, "LZ4 stream decompression failed");
+  DebugAssert(static_cast<size_t>(decompressed_result) == decompressed_block_size,
+              "Decompressed LZ4 block has different size than the initial source data.")
+}
+
+template <typename T>
+void LZ4Segment<T>::_decompress_block_to_bytes(const size_t block_index, std::vector<char>& decompressed_data) const {
+  // Assure that the decompressed data fits into the vector.
+  if (decompressed_data.size() != _block_size) {
+    decompressed_data.resize(_block_size);
+  }
+
+  // We use the string method since we handle a char-vector (even though the data is no necessarily string data).
+  _decompress_block_to_bytes(block_index, decompressed_data, 0u);
+
+  /**
+    * In the case of the last block, the decompressed data is possibly smaller than _block_size (it is _last_block_size
+    * large). However, when using _last_block_size as size for the decompression buffer, the LZ4 decompression fails.
+    * Therefore, the block is resized and shrunk afterwards.
```

Comment:
Where is it shrunk? And why would we want that?

Reply:
We removed the shrinking since the vector is re-used to store each decompressed block.